src/main/java/rx/internal/util/BackpressureDrainManager.java

```diff
@@ -0,0 +1,240 @@
+
+/**
+ * Manages the producer-backpressure-consumer interplay by
+ * matching up available elements with requested elements and/or
+ * terminal events. 
+ */
+@Experimental
+public final class BackpressureDrainManager implements Producer {
+    /**
+     * Interface representing the minimal callbacks required
+     * to operate the drain part of a backpressure system.
+     */
+    public interface BackpressureQueueCallback {
+        /**
+         * Override this method to peek for the next element,
+         * null meaning no next element available now.
+         * <p>It will be called plain and while holding this object's monitor.
+         * @return the next element or null if no next element available
+         */
+        Object peek();
+        /**
+         * Override this method to poll (consume) the next element,
+         * null meaning no next element available now.
+         * @return the next element or null if no next element available
+         */
+        Object poll();
+        /**
+         * Override this method to deliver an element to downstream.
+         * The logic ensures that this happens only in the right conditions.
+         * @param value the value to deliver, not null
+         * @return true indicates that one should terminate the emission loop unconditionally
+         * and not deliver any further elements or terminal events.
+         */
+        boolean accept(Object value);
+        /**
+         * Override this method to deliver a normal or exceptional
+         * terminal event.
+         * @param exception if not null, contains the terminal exception
+         */
+        void complete(Throwable exception);
+    }
+    
+    /** The request counter, updated via REQUESTED_COUNTER. */
+    protected volatile long requestedCount;
+    /** Atomically updates the the requestedCount field. */ 
+    protected static final AtomicLongFieldUpdater<BackpressureDrainManager> REQUESTED_COUNT
+        = AtomicLongFieldUpdater.newUpdater(BackpressureDrainManager.class, "requestedCount");
+    /** Indicates if one is in emitting phase, guarded by this. */
+    protected boolean emitting;
+    /** Indicates a terminal state. */
+    protected volatile boolean terminated;
+    /** Indicates an error state, barrier is provided via terminated. */
+    protected Throwable exception;
+    /** The callbacks to manage the drain. */
+    protected final BackpressureQueueCallback actual;
+    /**
+     * Constructs a backpressure drain manager with 0 requesedCount,
+     * no terminal event and not emitting.
+     * @param actual he queue callback to check for new element availability
+     */
+    public BackpressureDrainManager(BackpressureQueueCallback actual) {
+        this.actual = actual;
+    }
+    /**
+     * Checks if a terminal state has been reached.
+     * @return true if a terminal state has been reached
+     */
+    public final boolean isTerminated() {
+        return terminated;

```

Comment:
`isTerminated`, `terminate`, `terminate` are not used. Could you explain why adding them?